collector/user_input_delay.go

```diff
@@ -0,0 +1,135 @@
+			"Maximum value for queuing delay across all user input waiting to be picked-up by the process during a target time interval",
+			[]string{"session_id", "process_id", "process_name"},
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *UserInputCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ctx, ch); err != nil {
+		log.Error("failed collecting user_input_delay metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+// Perflib: "User Input Delay per Session"
+type SessionMaxInputDelay struct {
+	Name string
+
+	MaxInputDelay float64 `perflib:"MaxInputDelay"`
+}
+
+// Perflib: "User Input Delay per Process"
+type ProcessMaxInputDelay struct {
+	Name string
+
+	MaxInputDelay float64 `perflib:"MaxInputDelay"`
+}
+
+type processLabels struct {
+	sessionID   string
+	processID   string
+	processName string
+}
+
+// Process name provided by "User Input Delay per Process" ListSet can be split to multiple labels.
+// E.G. "1:7232 <TextInputHost.exe>" could be split to:
+//  sessionID: 1
+//  processID: 7232
+//  processName: TextInputHost.exe
+func splitProcessLabel(process_label string) (*processLabels, error) {
+	// Use regex in place of multiple splits
+	query := regexp.MustCompile(`(\d):(\d+)\s+<(.*)>`)
+	if !query.MatchString(process_label) {
+		return nil, errors.New(fmt.Sprint("Unexpected process label structure: ", process_label))
+	}
+	result := query.FindStringSubmatch(process_label)
+
+	labels := processLabels{
+		sessionID:   result[1],
+		processID:   result[2],
+		processName: result[3],
+	}
+
+	return &labels, nil
+}
+
+func (c *UserInputCollector) collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+	var session_dst []SessionMaxInputDelay
+	// Note that collector will exit early if Session CounterSet cannot be queried, resulting in the Process Couterset being skipped.
+	if err := unmarshalObject(ctx.perfObjects["User Input Delay per Session"], &session_dst); err != nil {
+		return nil, err
+	}
+
+	var process_dst []ProcessMaxInputDelay
+	if err := unmarshalObject(ctx.perfObjects["User Input Delay per Process"], &process_dst); err != nil {
+		return nil, err
+	}
+
+	for _, session := range session_dst {
+		// Skip Average and Max session IDs, these can be computed with Prometheus functions.
+		if session.Name == "Average" || session.Name == "Max" {
+			continue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.SessionMaxInputDelay,
+			prometheus.GaugeValue,
+			session.MaxInputDelay,
+			session.Name,
+		)
+	}
+
+	for _, process := range process_dst {
+		labels, err := splitProcessLabel(process.Name)
+		if err != nil {
+			return nil, err
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.ProcessMaxInputDelay,
+			prometheus.GaugeValue,
+			process.MaxInputDelay,
+			labels.sessionID,
+			labels.processID,
+			labels.processName,
```

Comment:
What sort of cardinality does this result in, btw? Something we need to warn the user about?

Reply:
Once of the labels is a process ID, so the cardinality will be quite high. I'll add a warning similar to the `process` and `service` collectors.

Reply:
Similar warning has been added, as well as the whitelist/blacklist flags from the `process` collector.